Source/ARTRealtime.m

```diff
@@ -381,21 +391,29 @@ - (void)transitionSideEffects:(ARTConnectionStateChange *)stateChange {
             }
         }
     }
+
+    return waitingForCurrentEventSemaphore;
 }
 
-- (void)unlessStateChangesBefore:(NSTimeInterval)deadline do:(void(^)())callback {
-    // Defer until next event loop execution so that any event emitted in the current
-    // one doesn't cancel the timeout.
+- (_Nonnull dispatch_semaphore_t)unlessStateChangesBefore:(NSTimeInterval)deadline do:(void(^)())callback __attribute__((warn_unused_result)) {
+    // Defer until next event loop execution so that any event emitted in the current one doesn't cancel the timeout.
     ARTRealtimeConnectionState state = self.connection.state;
-    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0), dispatch_get_main_queue(), ^{
+    // Timeout should be dispatched after current event.
+    dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0), _eventQueue, ^{
+        // Wait until the current event is done.
+        dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER);
+        ARTEventListener *listener = [_internalEventEmitter once:^(ARTConnectionStateChange *change) {
```

Comment:
Why do we always create the listener, ie. why the condition on line 409 has been moved below this?